### PR TITLE
Features/new structure

### PR DESCRIPTION
**Pull request by zakrad**

### CODE REVIEW

File: contracts/DLT/DLT.sol [hunk 1]
```diff
@@ -379,14 +381,10 @@ contract DLT is Context, IDLT {
...
 
-        unchecked {
-            _totalSupply += amount;
-            _mainTotalSupply[mainId] += amount;
-            _subTotalSupply[mainId][subId] += amount;
-
-            _balances[mainId][account].mainBalance += amount;
-            _balances[mainId][account].subBalances[subId] += amount;
-        }
+        _totalSupply += amount;
```
Comment (Aboudjem) [inline contracts/DLT/DLT.sol:384]: should we keep the total supply? is it required here?

File: contracts/DLT/DLT.sol [hunk 2]
```diff
@@ -379,14 +381,10 @@ contract DLT is Context, IDLT {
...
+        _mainTotalSupply[mainId] += amount;
+        _subTotalSupply[mainId][subId] += amount;
```
Comment (Aboudjem) [inline contracts/DLT/DLT.sol:386]: I think these totalSupply, can be retrieved from the `_balances`, do we really need to have 4 storage access here?

File: contracts/DLT/DLT.sol (hunk 1 above)
Comment (zakrad) [reply contracts/DLT/DLT.sol:384]: no

File: contracts/DLT/extensions/DLTEnumerable.sol [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+    function _burn(
+        address recipient,
+        uint256 mainId,
+        uint256 subId,
+        uint256 amount
+    ) internal virtual override(DLT) {
+        super._burn(recipient, mainId, subId, amount);
+        unchecked {
+            _mainTotalSupply[mainId] -= amount;
+            _subTotalSupply[mainId][subId] -= amount;
```
Comment (mohamadhammoud) [inline contracts/DLT/extensions/DLTEnumerable.sol:135]: Referring to openzeppeline library, they didn't go with overriding a _burn method in an Enumerable smart contract.
I would suggest if we can go with Enumerable functionality by _beforeTokenTransfer method.

It is the best choice as extension to add functionality from side methods not main ones

File: contracts/DLT/interfaces/IDLTEnumerable.sol [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+     */
+    function subIds(uint256 mainId) external view returns (uint256[] memory);
+
+    /**
+     * @dev Returns total sub id balance of owner for each main id
+     */
+    function totalSubIdBalance(
+        address owner,
+        uint256 mainId
+    ) external view returns (uint256);
```
Comment (mohamadhammoud) [inline contracts/DLT/interfaces/IDLTEnumerable.sol:44]: may we think for better names ? we can specify names that let the developer specify what does this method do from just the name.

File: contracts/DLT/extensions/DLTEnumerable.sol [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+        if (_subTotalSupply[mainId][subId] == 0) {
+            ++_totalSubIds[mainId];
+            uint256[] storage array = _subIds[mainId];
+            _subIdIndex[mainId][subId] = array.length;
+            array.push(subId);
+        }
+        _mainTotalSupply[mainId] += amount;
+        _subTotalSupply[mainId][subId] += amount;
+        super._mint(recipient, mainId, subId, amount);
+    }
```
Comment (mohamadhammoud) [inline contracts/DLT/extensions/DLTEnumerable.sol:120]: why do we need to mint from extension ?

**Review by Aboudjem — APPROVED** [review]

**Pull request merged**